Bridge polls

clehner opened the pull request:
Two-way bridging of Jitsi polls with IRC.

Polls are identified by a number starting at one. Options in each poll are also identified by a number starting at one.

## Poll creation

### From Jitsi
![Jitsi poll creation](https://demo.didkit.dev/2021/12/30/jitsipoll-create.png)
```
<cgbot-cel> Charles_E._Lehner created poll #1 with 2 options.
<cgbot-cel> ... Question: Why?
<cgbot-cel> ... Option 1: Because.
<cgbot-cel> ... Option 2: Why not?
```

### From IRC
```
<@cel> cgbot-cel: create poll Example | A1 | B2
<cgbot-cel> cel created a poll #2 with 2 options.
<cgbot-cel> POLL: Example
<cgbot-cel> ... Option 1: A1
<cgbot-cel> ... Option 2: B2
```
![Jitsi poll created via IRC](https://demo.didkit.dev/2022/01/08/jitsipoll.png)

## Voting in a poll

### From Jitsi
![voting-in-poll](https://demo.didkit.dev/2021/12/30/vote.png)
```
<cgbot-cel> Charles E. Lehner answered poll #1: (2) Why not?
```

### From IRC

```
<@cel> cgbot-cel: vote 1
<cgbot-cel> cel answered poll #1: (1) Because.
```
![votes](https://demo.didkit.dev/2021/12/30/votes.png)

![votes-expanded](https://demo.didkit.dev/2021/12/30/votes-details.png)

Multiple votes:
```
<@cel> cgbot-cel: vote 1,2
<cgbot-cel> cel answered poll #1: (1) Because., (2) Why not?
```
![votes-multi-expanded](https://demo.didkit.dev/2021/12/30/votes-multi-expanded.png)

### Querying poll state

Show info about the current poll or a particular poll:
```
<@cel> cgbot-cel: poll
<cgbot-cel> Poll #1 with 2 options.
<cgbot-cel> ... Question: Why?
<cgbot-cel> ... Option 1: Because.
<cgbot-cel> ... Option 2: Why not?
```

List all polls since the bot joined:
```
<@cel> cgbot-cel: polls
<cgbot-cel> 1 polls.
<cgbot-cel> ... Poll #1: Why?
```

Show votes on the most recent poll or a particular poll:
```
<@cel> cgbot-cel: votes
<cgbot-cel> Poll #1 with 2 options has 2 votes from 2 voters.
<cgbot-cel> ... Question: Why?
<cgbot-cel> ... Option 1 has 1 votes (50%); voted by: cel
<cgbot-cel> ... Option 2 has 1 votes (50%); voted by: Charles E. Lehner
```

clehner committed:
Bridge polls

msporny commented on lib/command.js:
```diff
@@ -1,10 +1,43 @@
 
-export async function processCommand({nick, message, participants, say}) {
+function nonCryptographicHash8(data) {
+  return crypto.createHash('md5').update(data).digest('hex').slice(0, 8);
```
Don't use known broken hash functions, even if md5, even if it's safe to do so in this particular instance. Other developers will come in behind you and copy what you did here, which might proliferate the bad md5 function elsewhere. Just do a sha-256 and take the last 8 characters, the result is the same, but without risking bad practice proliferation.

Granted, this entire file is bad code due to 10+ years of neglect; we need a rewrite at some point in the near future.

clehner replied:
Fixed: a19e6414da7e21f64d254ae3660384ce164635bd

msporny commented on lib/command.js:
```diff
@@ -1,10 +1,43 @@
+  if(args.length > 0) {
+    pollNum = args.shift().replace(/^#/, '');
+    if(isNaN(pollNum)) {
+      say('Expected poll number');
```
Remember that almost no one will remember the syntax. If you raise an error, tell people what a typical good command looks like.

clehner replied:
This part of the code (the `getPoll` function) can be reached from multiple commands (vote, votes, and poll).
The poll number is the last part to parse: for the `vote` command the vote argument has already been parsed at this point; for `poll` and `votes`, the poll is the only argument.

How's this? ade680cc2be41df1e56b36b07a4fb9deffbd3968
```
21:46 <@cel> cgbot-cel: vote 4 on \a
21:46 <cgbot-cel> Unrecognized poll "\a". Try e.g. "1", "#1", or "poll 1".
```

```
21:47 <@cel> cgbot-cel: poll asdf
21:47 <cgbot-cel> Unrecognized poll "asdf". Try e.g. "1", "#1", or "poll 1".
```
```
21:49 <@cel> cgbot-cel: votes 1!
21:49 <cgbot-cel> Unrecognized poll "1!". Try e.g. "1", "#1", or "poll 1".
```

msporny commented on lib/command.js:
```diff
@@ -1,10 +1,43 @@
+    }
+  } else {
+    if(polls.count === 0) {
+      say('No polls.');
```
Same as above, how does one create a poll if there are no polls?

clehner replied:
Added: 97561c9ade9c0b2513715a5b0b056964f9dfee1f, 183cf05bafe4e5cd472d4147807db14e75effd32, 621d5b2dd4f20f520b8b225f7e82a9e99dd34dd8, d90800210f69d64dda54fc6e7f7a01b8074dc9dd.

msporny commented on lib/command.js:
```diff
@@ -1,10 +1,43 @@
+    pollNum = polls.count;
+  }
+  if(args.length > 0) {
+    say('Unexpected data after poll number')
```
Same, show them what a correct command looks like.

clehner replied:
In this case the command can (maybe) be corrected by removing the unexpected trailing arguments. To output the whole (modified) command may be possible but not straightforward, because the command/arguments are parsed mutably (using `shift`). But the unrecognized part can be highlighted. How's this? bc189e3aaf425c12806f2b03be764c14e525cb76
```
22:06 <@cel> cgbot-cel: vote 1 on 4 a
22:06 <cgbot-cel> Unexpected data after poll number. Try again without this part: "a"
```

```
22:07 <@cel> cgbot-cel: poll 1 2
22:07 <cgbot-cel> Unexpected data after poll number. Try again without this part: "2"
```

msporny commented on lib/command.js:
```diff
@@ -1,10 +1,43 @@
+  }
+  const pollId = polls.idsByNum[pollNum];
+  if(!pollId) {
+    say(`Unknown poll "${pollNum}"`);
```
Might want to list known poll numbers here.

clehner replied:
Added: 3d477e502a809f9db86f61b8f6eb44621a4a6c6e, 621d5b2dd4f20f520b8b225f7e82a9e99dd34dd8, d90800210f69d64dda54fc6e7f7a01b8074dc9dd.

msporny commented on lib/command.js:
```diff
@@ -97,6 +130,152 @@ export async function processCommand({nick, message, participants, say}) {
     process.kill(process.pid, 'SIGINT');
     return;
   }
+  if(command == 'create' && args[0] === 'poll') {
```
We really need to factor all of these commands out into separate files... this file is getting large. Just noting this for now.

msporny commented on lib/command.js:
```diff
@@ -97,6 +130,152 @@ export async function processCommand({nick, message, participants, say}) {
+      if(args[0] === 'on') {
+        args.shift();
+      } else {
+        say('Try: vote option X on poll Y');
```
Might try making "on poll Y" optional and just use the last poll sent to the channel?

clehner replied:
That is the case. If "on poll Y" is omitted, the most recently created poll is used, and this highlighted code path is not reached.

msporny commented on lib/command.js:
```diff
@@ -97,6 +130,152 @@ export async function processCommand({nick, message, participants, say}) {
+      answersMap[answerI] = true;
+      const answer = poll.msg.answers[answerI];
+      if(answer == null) {
+        say(`Unknown option "${option}"`);
```
People might get confused here since IRC is async, so people's votes might not get counted and they might not realize?

clehner replied:
Do you mean in this particular error (unknown option chosen)? Or in all the error/failure cases for this command?
The responses could say something additional like "vote not processed, try again" to be more explicit about the vote not being counted? Or the response could be targeted to the user (maybe using `<nick>, message` since `nick: message` might get interpreted as scribing?) so they would be more likely to notice that it is relevant to them. There could also be unhandled vote attempts that are not matched at all by this handler and so would not generate a response. But a successful vote will result in a IRC message saying that the answer/vote took place, as generated by the bot when it receives back its own `answer-poll` xmpp message, and that (along with the vote appearing in the Jitsi UI) is the positive confirmation.

clehner replied:
Changed to list available options if an unknown option is picked: 2bb6c537a682506bb1573e458addd516208c706f
```
22:14 <@cel> cgbot-cel: create poll Test? | Yes | No
22:14 <cgbot-cel> cel created a poll #1 with 2 options.
22:14 <cgbot-cel> POLL: Test?
22:14 <cgbot-cel> ... Option 1: Yes
22:14 <cgbot-cel> ... Option 2: No
22:14 <@cel> cgbot-cel: vote 3
22:14 <cgbot-cel> Unknown option "3" on poll #1. Available options: 1,2
```

msporny commented on lib/manage.js:
```diff
@@ -165,22 +166,80 @@ export async function manage({meeting, meetings, xmppOptions}) {
+            });
+          }
+
+          // the message is a poll answer message
+          if(msg && msg.type === 'answer-poll') {
+            const poll = polls.byId[msg.pollId];
+            if(!poll) {
+              console.error('Unknown poll ' + msg.pollId);
+              return;
+            }
+            const answers = poll.msg.answers;
+            poll.votes[msg.voterId] = msg;
+            const message = `${msg.voterName} answered poll #${poll.num}: `
+              + msg.answers.map((chosen, i) => chosen
+                ? `(${i+1}) ${answers[i]}` : '')
+                .filter(Boolean).join(', ');
+            ircSay(message);
+            if(ircOptions.log) {
+              log({logFileBasePath, nick: ircOptions.nick, message});
```
Is there a "finalize poll" or "close poll" command that shows the totals? It might be difficult for more complex polls to understand the final tally? I know we have the "PROPOSAL / RESOLUTION" syntax already for proposals... but POLLs are different, maybe?

clehner replied:
The `votes` IRC command added in this PR shows the votes summary for a poll (i.e. the latest poll, or a previous one if referenced); it aims to present similar information as the web view in Jitsi (i.e. number of votes per option, names of voters, and percentage of voters who chose each option). In the web UI, the number of voters is shown in the "details" view but in this IRC implementation I have it show every time - If that is excessive with many voters, it could be adjusted.

There's no operation in the Jitsi front-end to close, remove or finalize a poll. New polls can be created and the old ones remain. The closest I can see that we could do to close a poll purely in the bot, I think would be the following:
- The bot could state in IRC and Jitsi chat that the poll is to be considered closed.
- The bot could zero out the poll - removing all existing votes. I think this could work because the web UI and conference system does not appear to verify that a vote message's sender property corresponds to the actual message sender; also, even though a user cannot remove their vote in the web UI (the client requires at least one option chosen), I don't think that is enforced internally.
- The bot could detect and revert votes made on a poll that has been marked closed. Additionally it could send a private message to the voter as a warning/alert.

Edit: actually, a poll can also be reset by sending a `new-poll` message with the existing poll id. That clears the votes., and can change the poll question and options. So, to close a poll, the options could be set to an empty list, and the title have e.g. " [CLOSED]" appended to it. It will still appear in the polls list but no one will able to vote in it.

Polls are different from proposal/resolutions but might be similarly useful in some situations.
For example, to emulate the current proposal method of evaluating a proposed resolution, a poll could have the question be a proposal, and options "+1", "0", and "-1" (or "yes", "no", "abstain", "block", "still deciding").

If there are multiple alternatives to be considered as proposals, a single poll could have a more general question, and give the proposed alternatives as options. Participants can vote for more than one option in a poll. This could be useful for comparing multiple possible alternatives. This might not have the nuance of the earlier method, as each option would get only a binary response per person (or ternary if you consider not voting as a response) rather than a spectrum of responses. It could be followed up with a poll specifically for an option to consider further.

msporny commented on lib/manage.js:
```diff
@@ -165,22 +166,80 @@ export async function manage({meeting, meetings, xmppOptions}) {
+            if(ircOptions.log) {
+              log({logFileBasePath, nick: ircOptions.nick, message});
+            }
+            const message2 = `... Question: ${msg.question}`;
```
```suggestion
            const message2 = `POLL: ${msg.question}`;
```

Let's do this to keep with the current syntax for "PROPOSAL/USE CASE/RESOLUTION/etc."

clehner replied:
Applied: f119cc316b687f204b8265c0d558c4a6c92d9302

msporny commented on lib/manage.js:
```diff
@@ -165,22 +166,80 @@ export async function manage({meeting, meetings, xmppOptions}) {
+            // bot, so use senderName instead.
+            const bridged = stanza.attrs.from.endsWith('/cgbot');
+            const name = bridged ? msg.senderName : nick;
+            const message = `${name} created poll #${pollNum} with ${numOptions} options.`;
```
```suggestion
            const message = `${name} created a poll #${pollNum} with ${numOptions} options.`;
```

clehner replied:
Applied: db3508bb90d4722ae9b988ee109f0be86919fd40

msporny reviewed (commented):
Thank you for the new feature, @clehner! Overall, LGTM. I'm assuming you've tested this -- remember that people get /super cranky/ during a call when the cgbot disappears/crashes. It's set up to re-join, but if there is some bug that happens when it rejoins and it crashes again, it stops the groups ability to scribe minutes... just putting that out there as a concern and to underscore that any addition to cgbot needs copious amounts of testing.

msporny reviewed (approved):
+1 to merge as long as comments are processed.

clehner committed:
Use sha256 for voter ids

clehner commented:
@msporny thanks for the review. I'll be addressing the things you pointed out. I believe I've tested (manually) all the added functionality so far. After making changes I'll test again and write what was tested.

clehner committed:
Use getPoll in vote command

clehner committed:
Prefix poll question with "POLL"

clehner committed:
Refer to "a" poll created

clehner committed:
Pass xmpp function to group chat command handler

clehner committed:
Fix indentation

clehner committed:
Pass bot nick to command processor

clehner committed:
Say how to create a poll when there are none

clehner committed:
List current polls when one is not matched

clehner committed:
Correct pluralization

clehner committed:
Show create command in response to polls command

clehner committed:
Detect empty polls list after unknown poll number

clehner committed:
Note polls can be created in Jitsi

clehner committed:
More output when unable to parse poll number

clehner committed:
Echo unexpected data after poll number

clehner committed:
List available options if unknown option chosen

clehner committed:
Add some comments

clehner committed:
Remove unused variable

clehner committed:
Set poll sender name

- Set bot sender id/name for new-poll message.
- Put chat nick id in question prefix and in separate message property.
- Strip question prefix when processing message.

clehner committed:
Use random string for new poll id

Prevent accidentally resetting poll by reusing id.

clehner committed:
Check question type

clehner committed:
Allow "in" instead of "on"

clehner committed:
Pass polls objects to command processor from XMPP

clehner committed:
Give command suggestion in polls command error

clehner commented:
@msporny
I've applied and/or responded to the requested changes.

### Other changes

Additionally I made some other minor changes, mostly for error handling; the most significant one is probably 38c132208694f3cb1f3c6c17ee68c0e621d0fc32 (Set poll sender name), described below:
<details>
<summary>38c132208694f3cb1f3c6c17ee68c0e621d0fc32</summary>

When creating a poll from chat (IRC/XMPP), the nick could not be set in the poll message; it would appear empty because the sender id was not a valid Jitsi chat participant. The fix implemented is to put the nick in the poll question (e.g. `<nick> Question`), and use the bot as the sender/name. The effect in Jitsi's web UI can be seen in the following screenshots.

#### Before
```
17:10 <@cel> cgbot-cel: create poll Example A | One | Two
17:10 <cgbot-cel> cel created a poll #1 with 2 options.
17:10 <cgbot-cel> POLL: Example A
17:10 <cgbot-cel> ... Option 1: One
17:10 <cgbot-cel> ... Option 2: Two
```
![before](https://demo.didkit.dev/2022/01/08/example-before-2.png)

#### After
```
17:11 <@cel> cgbot-cel: create poll Example B | One | Two
17:11 <cgbot-cel> cel created a poll #1 with 2 options.
17:11 <cgbot-cel> POLL: Example B
17:11 <cgbot-cel> ... Option 1: One
17:11 <cgbot-cel> ... Option 2: Two
```
![after](https://demo.didkit.dev/2022/01/08/example-after-1.png)
</details>

### Documentation of IRC commands

The bot has a help command that responds with a link to http://bit.ly/2CR6pZK, which is a redirect to <https://github.com/w3c-ccg/w3c-ccg.github.io/blob/master/irc_ref.md>. I opened a PR to add documentation to that page for the poll commands as currently implemented: https://github.com/w3c-ccg/w3c-ccg.github.io/pull/27

### Testing
I tested the commands and operations in various scenarios, as follows.
<details>
<summary>Tests</summary>

- [x] Create poll in Jitsi
  - [x] Poll creation is shown in IRC
  - [x] Create poll in web client
  - [x] Create poll in mobile app
- [x] Vote in poll in Jitsi
  - [x] Vote in web client
  - [x] Vote in mobile app
  - [x] Vote is shown in IRC
  - [x] Initial vote
  - [x] Change vote
  - [x] Single choice
  - [x] Multiple choices
  - [x] Vote in current poll
  - [x] Vote in previous poll
- [x] Vote in poll with "vote" command
  - [x] Vote in IRC
  - [x] Vote in Jitsi chat (Note: does not result in chat reply on success)
  - [x] Vote is shown in Jitsi web app
  - [x] Vote is shown in Jitsi mobile app
  - [x] Vote is shown in IRC
  - [x] Initial vote
  - [x] Change vote
  - [x] Single choice
  - [x] Multiple choices
  - [x] Vote in current poll
  - [x] Vote in previous poll
  - [x] No polls
  - [x] Bad option number
  - [x] Bad poll number format
  - [x] Unknown poll
  - [x] Unexpected additional arguments
- [x] Listing polls with "polls" command
  - [x] No polls
  - [x] One poll
  - [x] Two polls
  - [x] Unexpected additional arguments
- [x] Querying poll with "poll" command
  - [x] From IRC
  - [x] From Jitsi chat
  - [x] No polls
  - [x] Current poll
  - [x] Previous poll
  - [x] Bad poll number format
  - [x] Unknown poll
  - [x] Unexpected additional arguments
- [x] Listing votes with "votes" command
  - [x] From IRC
  - [x] From Jitsi chat
  - [x] No polls
  - [x] Current poll
  - [x] Previous poll
  - [x] Bad poll number format
  - [x] Unknown poll
  - [x] Unexpected additional arguments
  - [x] From IRC
  - [x] From Jitsi chat
</details>

msporny commented:
Oh, no... did I never merge this!? I thought I merged this a while ago... just noticed as I went to look at the latest code base for cgbot. Looks like we need to do some merging... I'll try to get mainline fixed first... then merge this on top?

clehner committed:
Merge remote-tracking branch 'origin/main' into feat/polls

clehner commented:
OK

I fixed some merge conflicts just now.
